src/node/blockstorage.cpp

```diff
@@ -785,19 +785,27 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, c
     return true;
 }
 
-/** Store block on disk. If dbp is non-nullptr, the file is known to already reside on disk */
 FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, CChain& active_chain, const CChainParams& chainparams, const FlatFilePos* dbp)
 {
     unsigned int nBlockSize = ::GetSerializeSize(block, CLIENT_VERSION);
     FlatFilePos blockPos;
-    if (dbp != nullptr) {
+    const auto position_known {dbp != nullptr}; // if dbp is not nullptr, then the file and block within it are known to already reside on disk
+    if (position_known) {
         blockPos = *dbp;
-    }
-    if (!FindBlockPos(blockPos, nBlockSize + 8, nHeight, active_chain, block.GetBlockTime(), dbp != nullptr)) {
+    } else {
+        // when known, blockPos.nPos points at the offset of the block data in the blk file. that already accounts for
+        // the serialization header present in the file (the 4 magic message start bytes + the 4 length bytes = 8 bytes).
+        // we add 8 only for new blocks since they will have the serialization header added when written to disk.
+        // this avoids over accounting for the serialization header on existing blocks. such over accounting was a long
+        // standing bug that added undesirable 8 byte gaps into blk data files following a -reindex operation.
+        // for more info, see https://github.com/bitcoin/bitcoin/issues/21379
+        nBlockSize += 8U;
```

Comment:
Could also be phrased as `nBlockSize += sizeof(unsigned int) + CMessageHeader::MESSAGE_START_SIZE` to avoid hardcoding the expected header size.

We could also include 
```cpp
constexpr size_t BLOCK_SERIALIZATION_HEADER_LEN = 
    sizeof(unsigned int) + CMessageHeader::MESSAGE_START_SIZE;
```
nearby `WriteBlockToDisk` to avoid having that information live too far away from its source of truth.

Reply:
maybe. this actually crossed my mind. i wasn't sure if it was better to go with the hardcoded 8 that has been around for so long, or to go with the sizeof(unsigned int) that could be wrong for LP32 (https://en.cppreference.com/w/cpp/language/types). we may go with what you suggest, but i'd like to get some feedback and make this change on purpose.

Reply:
i just found in src/compat/assumptions.h `static_assert(sizeof(unsigned) == 4, "32-bit unsigned assumed");`
that is what i need to feel safe making this change. i'll do it in my next update.

Reply:
update in 425fc8813307b8d96a48e4b45d0a7b00bbc2289e